vc-got.el

```diff
@@ -568,5 +568,64 @@ LIMIT limits the number of commits, optionally starting at START-REVISION."
                  (vc-got--diff file)))
               (t (error "Not implemented")))))))
 
+(defun vc-got-annotate-command (file buf &optional rev)
+  "Show annotated contents of FILE in buffer BUF. If given, use revision REV."
+  (let (process-file-side-effects)
+    (with-current-buffer buf
+      (apply #'vc-got--call "blame" (if rev
+                                        (list "-c" rev file)
+                                      (list file))))))
+
+;; captures 1=REV, 2=DATE, 3=USER
+(defconst vc-got-annotate-re
+  (concat "^[0-9]\\{1,\\}) " ; line number followed by )
+          "\\([a-z0-9]+\\) " ; SHA-1 of commit
+          "\\([0-9]\\{4\\}-[0-9]\\{2\\}-[0-9]\\{2\\}\\) "
+          "\\([^ ]\\)+ ")) ; year-mm-dd
+
+(defun vc-got-annotate-time ()
+  "Return the time of the next line of annotation at or after point.
+Value is returned as floating point fractional number of days."
+  (save-excursion
+    (beginning-of-line)
+    (when (looking-at vc-got-annotate-re)
+      (let ((str (match-string-no-properties 2)))
+        (vc-annotate-convert-time
+         (encode-time 0 0 0
+                      (string-to-number (substring str 8 10))
+                      (string-to-number (substring str 5 7))
+                      (string-to-number (substring str 0 4))))))))
+
+(defun vc-got-annotate-extract-revision-at-line ()
+  "Returns revision corresponding to the current line or nil."
+  (save-excursion
+    (beginning-of-line)
+    (when (looking-at vc-got-annotate-re)
+      (match-string-no-properties 1))))
+
+(defun vc-got-previous-revision (file rev)
+  "Return the revision number that precedes REV for FILE, or nil if no such revision exists."
+  (with-temp-buffer
+    (vc-got--log file 2 rev nil nil t)
+    (goto-char (point-min))
+    (keep-lines "^commit")
+    (when (looking-at "^commit \\([a-z0-9]+\\)")
+      (match-string-no-properties 1))))
+
+(defun vc-got-next-revision (file rev)
+  "Return the revision number that follows REV for FILE, or nil
+  if no such revision exists."
+  (with-temp-buffer
+    (vc-got--log file nil nil rev)
+    (keep-lines "^commit" (point-min) (point-max))
+    (goto-char (point-max))
+    (forward-line -1) ;; return from empty line to last actual commit
+    (when (looking-at  "^commit \\([a-z0-9]+\\)")
+      ;; no need to continue if looking at top commit
+      (unless (string= rev (match-string-no-properties 1))
+        (forward-line -1)
+        (when (looking-at  "^commit \\([a-z0-9]+\\)")
+          (match-string-no-properties 1))))))
```

Comment:
The first `looking-at` is redundant, because we're guaranteed to be in a line that starts with "commit" followed by a SHA1.  Actually, we're also sure that that specific SHA1 is exactly equal to the `rev` parameter, because we're asking `got log` to stop at commit with sha1 `rev`.  So the unless check should always be true.

We need to retrieve the commit id in the previous line (if it exists).  We can do that replacing the whole `when` with something like

```elisp
(unless (= (point) (point-min))
  (forward-line -1)
  (when (looking-at "^commit \\([a-z0-9]+\\)")
    (match-string-no-properties 1)))
```

forward-line will place the cursor at the beginning of the line, so we know that there is a previous line by checking that we're not at `(point-min)`.  If that's the case, it goes up one line and extract the commit id.  To be fair, here the `when` isn't needed: we called `keep-lines` before, so every line starts with "commit", but it reads nice IMHO.  You can drop the when if you prefer.

With this, I believe all the annotate commands are done?  Looking forward to merge this!

By the way, I've never used the vc-annotate command before today, but it's amazing to press `n` and `p` in that buffer and navigate in the history of the changes!

Reply:
Ops, I forgot one thing.  Can you please add a `(require 'vc-annotate)`?  It fixes an error of the byte compiler about an unknown `vc-annotate-convert-time`.  Thanks!